File: 121.best_time_to_buy_and_sell_stock.py
import numpy as np
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        pmin = 1e4
        dmax = 0

        for i in range(len(prices)):

            if prices[i]-pmin > dmax:
                dmax = prices[i]-pmin

            elif prices[i] < pmin:
                pmin = prices[i]
        
        return dmax

File: 121.best_time_to_buy_and_sell_stock.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int min_price = prices[0];
        int max_profit = 0;
        for (int pi:prices) { 
            if (pi -  min_price > max_profit)
                max_profit = pi - min_price;
            else if (pi < min_price)
                min_price = pi;
        }
        return max_profit;
    }
};